tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/LockFreeExponentiallyDecayingReservoir.java

```diff
@@ -0,0 +1,218 @@
+ * Licensed under the Apache License, Version 2.0. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+@Beta
+public final class LockFreeExponentiallyDecayingReservoir implements Reservoir {
+
+    private static final AtomicReferenceFieldUpdater<LockFreeExponentiallyDecayingReservoir, State> stateUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(LockFreeExponentiallyDecayingReservoir.class, State.class, "state");
+
+    private final double alpha;
+    private final int size;
+    private final long rescaleThresholdNanos;
+    private final Clock clock;
+
+    private volatile State state;
+
+    private final class State {
+        private final long startTick;
+        private final AtomicLong count;
+        private final ConcurrentSkipListMap<Double, WeightedSample> values;
+
+        State(long startTick, AtomicLong count, ConcurrentSkipListMap<Double, WeightedSample> values) {
+            this.startTick = startTick;
+            this.values = values;
+            this.count = count;
+        }
+
+        private void update(long value, long timestampNanos) {
+            double itemWeight = weight(timestampNanos - startTick);
+            WeightedSample sample = new WeightedSample(value, itemWeight);
+            double priority = itemWeight / ThreadLocalRandom.current().nextDouble();
+
+            long newCount = count.incrementAndGet();
+            if (newCount <= size || values.isEmpty()) {
+                values.put(priority, sample);
+            } else {
+                Double first = values.firstKey();
+                if (first < priority && values.putIfAbsent(priority, sample) == null) {
+                    // Always remove an item
+                    while (values.remove(first) == null) {
+                        first = values.firstKey();
+                    }
+                }
+            }
+        }
+
+        State rescale(long newTick) {
+            long durationNanos = newTick - startTick;
+            double durationSeconds = durationNanos / 1_000_000_000D;
+            double scalingFactor = Math.exp(-alpha * durationSeconds);
+            AtomicLong newCount = new AtomicLong();
+            ConcurrentSkipListMap<Double, WeightedSample> newValues = new ConcurrentSkipListMap<>();
+            if (Double.compare(scalingFactor, 0) != 0) {
+                values.forEach((key, sample) -> {
+                    double newWeight = sample.weight * scalingFactor;
+                    if (Double.compare(newWeight, 0) == 0) {
+                        return;
+                    }
+                    WeightedSample newSample = new WeightedSample(sample.value, newWeight);
+                    if (newValues.put(key * scalingFactor, newSample) == null) {
+                        newCount.incrementAndGet();
+                    }
+                });
+            }
+            return new State(newTick, newCount, newValues);
```

Comment:
```suggestion
            return new State(newTick, new AtomicLong(newCount), newValues);
```

Reply:
Updated using a concrete class instead of a lambda in forEach to produce this result.